### DISCUSSION TITLE
Failed to install PyMuPDF version 1.19.4

### DISCUSSION THREAD

**Discussion by sunov** [pymupdf/PyMuPDF#1619] [General]
Hi Everyone, While installing PyMuPDF version 1.19.4, Getting following error. Could you please help anyone.

Python Version = 3.7.3

Building wheel for PyMuPDF (setup.py): started
  Building wheel for PyMuPDF (setup.py): finished with status 'error'
  ERROR: Complete output from command /usr/local/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-80ik6m9u/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-q9eukmre --python-tag cp37:
  ERROR: running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.7
  creating build/lib.linux-x86_64-3.7/fitz
  copying fitz/__init__.py -> build/lib.linux-x86_64-3.7/fitz
  copying fitz/fitz.py -> build/lib.linux-x86_64-3.7/fitz
  copying fitz/utils.py -> build/lib.linux-x86_64-3.7/fitz
  copying fitz/__main__.py -> build/lib.linux-x86_64-3.7/fitz
  running build_ext
  building 'fitz._fitz' extension
  creating build/temp.linux-x86_64-3.7
  creating build/temp.linux-x86_64-3.7/fitz
  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -Imupdf/thirdparty/freetype/include -I/usr/include/freetype2 -I/usr/local/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-x86_64-3.7/fitz/fitz_wrap.o
  fitz/fitz_wrap.c:2755:18: fatal error: fitz.h: No such file or directory
   #include <fitz.h>
                    ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for PyMuPDF
  Running setup.py clean for PyMuPDF
  Building wheel for PyYAML (setup.py): started
  Building wheel for PyYAML (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/16/27/a1/775c62ddea7bfa62324fd1f65847ed31c55dadb6051481ba3f
  Building wheel for pysmb (setup.py): started
  Building wheel for pysmb (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/10/3b/78/3e9614b1fdd80e63c6e0087056641559cc6dc96d5f14610bcf
  Building wheel for pycountry (setup.py): started
  Building wheel for pycountry (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/95/98/2e/d24621de79da77006d0c34553a2675f6b7ac9c061b8b2009dd
  Building wheel for compressed-rtf (setup.py): started
  Building wheel for compressed-rtf (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/3c/f5/de/f9bb69be4449e8d1745158f97e51aeba501080e62922e93922
  Building wheel for olefile (setup.py): started
  Building wheel for olefile (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/4b/f4/11/bc4166107c27f07fd7bba707ffcb439619197638a1ac986df3
  Building wheel for MarkupSafe (setup.py): started
  Building wheel for MarkupSafe (setup.py): finished with status 'done'
  Stored in directory: /root/.cache/pip/wheels/57/5b/c9/18e0a3bca98e7cfaec446b1c335405791168d5ac46553fdee5
Successfully built PyYAML pysmb pycountry compressed-rtf olefile MarkupSafe
Failed to build PyMuPDF
Installing collected packages: vine, amqp, Click, numpy, textdistance, scipy, tenacity, pyasn1, pysmb, pycountry, six, python-dateutil, pytz, pandas, etmfa-core-gxp, compressed-rtf, olefile, backports.zoneinfo, tzdata, pytz-deprecation-shim, tzlocal, imapclient, ebcdic, extract-msg, Werkzeug, MarkupSafe, Jinja2, itsdangerous, Flask, zipp, importlib-resources, attrs, typing-extensions, importlib-metadata, pyrsistent, jsonschema, aniso8601, flask-restplus, greenlet, gevent, joblib, kombu, pillow, PyMuPDF, pyxlsb, atomicwrites, more-itertools, wcwidth, py, pyparsing, packaging, pluggy, pytest, PyYAML, limits, pylogbeat, python-logstash-async, idna, chardet, urllib3, certifi, requests, xlrd, etmfa-triage
  Running setup.py install for PyMuPDF: started
    Running setup.py install for PyMuPDF: finished with status 'error'
    ERROR: Complete output from command /usr/local/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-80ik6m9u/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-5p_8xeb1/install-record.txt --single-version-externally-managed --compile:
    ERROR: running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/fitz
    copying fitz/__init__.py -> build/lib.linux-x86_64-3.7/fitz
    copying fitz/fitz.py -> build/lib.linux-x86_64-3.7/fitz
    copying fitz/utils.py -> build/lib.linux-x86_64-3.7/fitz
    copying fitz/__main__.py -> build/lib.linux-x86_64-3.7/fitz
    running build_ext
    building 'fitz._fitz' extension
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/fitz
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -Imupdf/thirdparty/freetype/include -I/usr/include/freetype2 -I/usr/local/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-x86_64-3.7/fitz/fitz_wrap.o
    fitz/fitz_wrap.c:2755:18: fatal error: fitz.h: No such file or directory
     #include <fitz.h>
                      ^
    compilation terminated.
    error: command 'gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command "/usr/local/bin/python -u -c 'import setuptools, tokenize;__file__='"'"'/tmp/pip-install-80ik6m9u/PyMuPDF/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-5p_8xeb1/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-80ik6m9u/PyMuPDF/

**Comment by JorjMcKie** [comment]
You are trying to install PyMuPDF from sources (... on purpose? There should be a wheel for your environment).
Anyway, you do not have installed MuPDF, a prerequisite for this.
See [this](https://github.com/pymupdf/PyMuPDF/tree/master/installation/ubuntu) installation procedure as an eample.

**Comment by sunov** [comment]
@JorjMcKie  We are calling PyMUPDF version from requirements.txt file without MuPDF and wheel file(pip install -e .), It is working fine for PyMUPDF==1.16.7 version. 

In same project, We are modifying requirements.txt file with PyMUPDF==1.19.4 version. It is creating above issue.

Collecting PyMuPDF==1.19.4
  Downloading PyMuPDF-1.19.4-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (8.7 MB)
While creating docker image, It is is taking linux wheel file automatically, Not creating any issue.

Collecting PyMuPDF==1.19.4 (from etmfa-triage==3.1)
  Downloading https://files.pythonhosted.org/packages/3b/7c/7d2d98a3f851a1bfc11b7bbeb345deb51a28a42ee7ace3244ce82f6c257f/PyMuPDF-1.19.4.tar.gz (2.3MB)

While running unit test cases trying to take wheel file from pythonhosted, And throwing above error.

**Comment by JorjMcKie** [comment]
Whatever the environment established by docker, it is not regarded as being compatible with any of the platform tags of the wheels on PyPI ... **_as per the perception of the used pip version._** This is why pip starts building from sources.
And the platform tags generated by `cibuildwheel` are a moving target. So for installing PyMuPDF, there is the recommendation to always upgrade pip first.
Don't know if this helps.
PyMuPDF at least does not have any external requirements. This has not changed ever.

**Reply by sunov** [reply]
@JorjMcKie By upgrading pip version, Worked for me. Thank you very much
